swarm/network/pss.go

```diff
@@ -1,52 +1,384 @@
+	}
+
 	return nil
 }
+
+// we need the pss pointer here to access the final send
+type PssReadWriter struct {
+	*Pss
+	RecipientOAddr pot.Address
+	LastActive     time.Time
+	rw             chan p2p.Msg
+	ct             *protocols.CodeMap
+	topic          *PssTopic
+}
+
+// get incoming msg
+func (prw PssReadWriter) ReadMsg() (p2p.Msg, error) {
+	msg := <-prw.rw
+
+	glog.V(logger.Detail).Infof("pssrw readmsg %v", msg)
+
+	return msg, nil
+}
+
+func (prw PssReadWriter) WriteMsg(msg p2p.Msg) error {
+	glog.V(logger.Detail).Infof("pssrw writemsg %v", msg)
+	ifc, found := prw.ct.GetInterface(msg.Code)
+	if !found {
+		return fmt.Errorf("Writemsg couldn't find matching interface for code %d", msg.Code)
+	}
+	msg.Decode(ifc)
+
+	to := prw.RecipientOAddr.Bytes()
+
+	glog.V(logger.Detail).Infof("pssrw writemsg %v", msg)
+
+	pmsg, _ := MakeMsg(msg.Code, ifc)
+
+	return prw.Pss.Send(to, *prw.topic, pmsg)
+}
+
+// insert a incoming message on the p2p.VirtualPeer hack
+func (prw PssReadWriter) injectMsg(msg p2p.Msg) error {
+	glog.V(logger.Detail).Infof("pssrw injectmsg %v", msg)
+	prw.rw <- msg
+	return nil
+}
+
+// encapsulates all the complexity of the bridging of real-to-virtual protocol instantiations
+// TODO: needs review to see if it can be made leaner
+type PssProtocol struct {
+	*Pss
+	virtualProtocol *p2p.Protocol
+	topic           *PssTopic
+	ct              *protocols.CodeMap
+}
+
+func NewPssProtocol(pss *Pss, topic *PssTopic, ct *protocols.CodeMap, targetprotocol *p2p.Protocol) *PssProtocol {
+	pp := &PssProtocol{
+		Pss:             pss,
+		virtualProtocol: targetprotocol,
+		topic:           topic,
+		ct:              ct,
+	}
+	return pp
+}
+
+func (self *PssProtocol) GetHandler() func([]byte, *p2p.Peer, []byte) error {
+	return self.handle
+}
+
+func (self *PssProtocol) handle(msg []byte, p *p2p.Peer, senderAddr []byte) error {
+	hashoaddr := pot.NewHashAddressFromBytes(senderAddr).Address
+	if !self.isActive(hashoaddr, *self.topic) {
+		rw := &PssReadWriter{
+			Pss:            self.Pss,
+			RecipientOAddr: hashoaddr,
+			rw:             make(chan p2p.Msg),
+			ct:             self.ct,
+			topic:          self.topic,
+		}
+		self.Pss.AddPeerTopic(hashoaddr, *self.topic, rw)
+		go func() {
+			err := self.virtualProtocol.Run(p, rw)
+			glog.V(logger.Detail).Infof("vprotocol on addr %v topic %v quit: %v", hashoaddr, self.topic, err)
+		}()
+		// we need a protocol ready reporting channel here
+	}
+
+	payload := &pssPayload{}
+	rlp.DecodeBytes(msg, payload)
+
+	pmsg := p2p.Msg{
+		Code:       payload.Code,
+		Size:       uint32(len(payload.Data)),
+		ReceivedAt: time.Now(),
+		Payload:    bytes.NewBuffer(payload.Data),
+	}
+
+	self.Pss.PeerPool[hashoaddr][*self.topic].injectMsg(pmsg)
+
+	return nil
+}
+
+func (ps *Pss) isSelfRecipient(msg *PssMsg) bool {
+	if bytes.Equal(ps.GetMsgRecipient(msg), ps.Overlay.GetAddr().OverlayAddr()) {
+		return true
+	}
+	return false
+}
+
+func (ps *Pss) GetMsgRecipient(msg *PssMsg) []byte {
```

Comment:
Because we might want to only reveal PART of the To-address in the unencrypted part of the PssMsg, and then the method for getting the real to-address would require more than just returning the member.